geo/src/algorithm/scale.rs

```diff
@@ -1,49 +1,150 @@
-///
-/// assert_eq!(scaled, line_string![
-///     (x: -45.0f64, y: -45.0f64),
-///     (x: 55.0f64, y: 55.0f64)
-/// ]);
-/// ```
-///
-pub trait Scale<T> {
-    fn scale(&self, x: T, y: T, origin: Point<T>) -> Self
-    where
-        T: CoordNum;
+pub trait Scale<T: CoordNum> {
+    /// Scale a geometry from it's bounding box center.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::Scale;
+    /// use geo::{LineString, line_string};
+    ///
+    /// let ls: LineString = line_string![(x: 0., y: 0.), (x: 10., y: 10.)];
+    ///
+    /// let scaled = ls.scale(2.);
+    ///
+    /// assert_eq!(scaled, line_string![
+    ///     (x: -5., y: -5.),
+    ///     (x: 15., y: 15.)
+    /// ]);
+    /// ```
+    #[must_use]
+    fn scale(&self, scale_factor: T) -> Self;
+
+    /// Mutable version of [`scale`](Self::scale)
+    fn scale_mut(&mut self, scale_factor: T);
+
+    /// Scale a geometry from it's bounding box center, using different values for `x_factor` and
+    /// `y_factor` to distort the geometry's [aspect ratio](https://en.wikipedia.org/wiki/Aspect_ratio).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::Scale;
+    /// use geo::{LineString, line_string};
+    ///
+    /// let ls: LineString = line_string![(x: 0., y: 0.), (x: 10., y: 10.)];
+    ///
+    /// let scaled = ls.scale_xy(2., 4.);
+    ///
+    /// assert_eq!(scaled, line_string![
+    ///     (x: -5., y: -15.),
+    ///     (x: 15., y: 25.)
+    /// ]);
+    /// ```
+    #[must_use]
+    fn scale_xy(&self, x_factor: T, y_factor: T) -> Self;
+
+    /// Mutable version of [`scale_xy`](Self::scale_xy).
+    fn scale_xy_mut(&mut self, x_factor: T, y_factor: T);
+
+    /// Scale a geometry around a point of `origin`.
+    ///
+    /// The point of origin is *usually* given as the 2D bounding box centre of the geometry, in
+    /// which case you can just use [`scale`](Self::scale) or [`scale_xy`](Self::scale_xy), but
+    /// this method allows you to specify any point.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::Scale;
+    /// use geo::{LineString, line_string};
+    ///
+    /// let ls: LineString = line_string![(x: 0., y: 0.), (x: 10., y: 10.)];
+    ///
+    /// let scaled = ls.scale_xy(2., 4.);
+    ///
+    /// assert_eq!(scaled, line_string![
+    ///     (x: -5., y: -15.),
+    ///     (x: 15., y: 25.)
+    /// ]);
+    /// ```
+    #[must_use]
+    fn scale_around_point(
+        &self,
+        x_factor: T,
+        y_factor: T,
+        origin: impl Into<Coordinate<T>>,
+    ) -> Self;
+
+    /// Mutable version of [`scale_around_point`](Self::scale_around_point).
+    fn scale_around_point_mut(
+        &mut self,
+        x_factor: T,
+        y_factor: T,
+        origin: impl Into<Coordinate<T>>,
+    );
 }
 
-impl<T, G> Scale<T> for G
+impl<T, IR, G> Scale<T> for G
 where
-    T: CoordNum,
-    G: MapCoords<T, T, Output = G> + MapCoordsInPlace<T>,
+    T: CoordFloat,
```

Comment:
Another potentially controversial decision: The new scale methods I've added rely on Rect::center which is only defined for Floats, so I opted to make this entire trait for Floats.

The alternative would be to split out the `scale_around_point` method into another trait, but I was afraid the ergonomic damage of having two different Scale traits wouldn't be worth it, and I do suspect that most people will want to do scaling from the center.

People who need to do integer scaling can still use `integer_geom.affine_transform(&AffineTransform::scale(2, 2))`

Reply:
Yep, I think this is fine, but I should also note that I'm biased here because [unintelligible unhelpful comments about integer geom functionality]